lnbits/core/services.py

```diff
@@ -23,8 +23,10 @@ def create_invoice(*, wallet_id: str, amount: int, memo: str) -> Tuple[str, str]
 
 
 def pay_invoice(*, wallet_id: str, bolt11: str, max_sat: Optional[int] = None) -> str:
-    temp_id = f"temp_{urlsafe_short_hash()}"
+    wallet = get_wallet(wallet_id)
+    assert wallet, "invalid wallet id"
```

Comment:
From the usage of this function I inferred that the wallet id is always valid at this point, so giving this function an invalid id should raise an error. I moved the check forward to avoid adding dummy payments to the wallet that wouldn't be removed due to an early exit of the function on failure.

Reply:
The problem with this is that we need to call `get_wallet().balance` as late as possible to avoid parallel payments going through when balance is not enough. Two options here:

- call again `get_wallet()` when checking the balance, so the temp payment is also taken into account (we make 2 queries then)
- move the get_wallet and assert lower. the temp payment will be created in any case, but it would be deleted after the exception has been catched

I added the temp payments to avoid sending different payments while we wait for the response from the backend. If we have another solution to avoid this problem, and it ius better, we can move things around again.

Reply:
Oh, I totally missed that, I kinda assumed something ORM-like to take care of fetching up to date values, but I remember seeing the raw queries. Will fix.